### ISSUE TITLE
Aliased `staticmethod` leads to incorrect attribute assignment

### ISSUE THREAD

**Issue by sharkdp** [microsoft/pyright#9820]
**Describe the bug**

In the following example, the aliased `staticmethod` decorator is not recognized properly and leads Pyright to believe that `C` instances have an attribute `x`.

While searching for existing issues, I found [this ticket](https://github.com/microsoft/pyright/issues/9108) where it is explained that the *"`@staticmethod` and `@classmethod` decorators […] require significant special casing in Python static type checkers"*. I understand that this might be difficult to implement. I do not rely on this behavior and just found it while testing edge-cases. Feel free to close this ticket without comment if this is out of scope for Pyright.

**Code or Screenshots**

```py
my_staticmethod = staticmethod

class Unrelated:
    x: int | None

class C:
    @my_staticmethod
    def f(unrelated: Unrelated):
        unrelated.x = 1

reveal_type(C().x)  # Pyright reports `int` here, but this attribute does not exist
```
([Open in Playground](https://pyright-play.net/?code=LYTw%2BgzgLghlCWBjYBTKALA9gEwAQF5do4lUMcAoCxAGxgglwFUA7AJxTqhWwC4Lcg3AA9eueCyi4APrgBymFiiq16jAML8huAAKhIsBMjRZsAodhQAzXFYAUAV3ac4PMaw5ceASi3ahTp6u2AB0wgS4AIxUHABuKDA0YFAgAA4odup23mHeggDEuAAKIGzwAOboUhypmGxQjAAGElCNuOgoHAA0uABGDlIY8IxwUGX93LjYmCiMLJhSKMLDUBRAA))

**VS Code extension or command-line**

Pyright 1.1.393

**Comment by erictraut** [comment]
This is expected behavior. Static type checkers like pyright and mypy do not generally handle aliases to `staticmethod` or `classmethod`. These decorators need to be processed early in the analysis — during the "binder" phase — when symbol tables are being constructed. This phase occurs prior to any type evaluation, which depends on the symbol tables.

If you want your code to work with static analysis tools, you'll need to avoid aliasing `staticmethod` and `classmethod`.
